src/particle.cpp

```diff
@@ -0,0 +1,681 @@
+      model::surfaces[surf_p->i_periodic_].get());
+
+    // Adjust the particle's location and direction.
+    bool rotational = other->periodic_translate(surf_p, this->r(), this->u());
+
+    // Reassign particle's surface
+    // TODO: off-by-one
+    surface_ = rotational ?
+      surf_p->i_periodic_ + 1 :
+      std::copysign(surf_p->i_periodic_ + 1, surface_);
+
+    // Figure out what cell particle is in now
+    n_coord_ = 1;
+
+    if (!find_cell(this, true)) {
+      this->mark_as_lost("Couldn't find particle after hitting periodic "
+        "boundary on surface " + std::to_string(surf->id_) + ".");
+      return;
+    }
+
+    // Set previous coordinate going slightly past surface crossing
+    r_last_current_ = this->r() + TINY_BIT*this->u();
+
+    // Diagnostic message
+    if (settings::verbosity >= 10 || simulation::trace) {
+      write_message("    Hit periodic boundary on surface " +
+        std::to_string(surf->id_));
+    }
+    return;
+  }
+
+  // ==========================================================================
+  // SEARCH NEIGHBOR LISTS FOR NEXT CELL
+
+#ifdef DAGMC
+  if (settings::dagmc) {
+    auto cellp = dynamic_cast<DAGCell*>(model::cells[cell_last_[0]].get());
+    // TODO: off-by-one
+    auto surfp = dynamic_cast<DAGSurface*>(model::surfaces[std::abs(surface_) - 1].get());
+    int32_t i_cell = next_cell(cellp, surfp) - 1;
+    // save material and temp
+    material_last_ = material_;
+    sqrtkT_last_ = sqrtkT_;
+    // set new cell value
+    coord_[0].cell = i_cell;
+    cell_instance_ = 0;
+    material_ = model::cells[i_cell]->material_[0];
+    sqrtkT_ = model::cells[i_cell]->sqrtkT_[0];
+    return;
+  }
+#endif
+
+  if (find_cell(this, true)) return;
+
+  // ==========================================================================
+  // COULDN'T FIND PARTICLE IN NEIGHBORING CELLS, SEARCH ALL CELLS
+
+  // Remove lower coordinate levels and assignment of surface
+  surface_ = 0;
+  n_coord_ = 1;
+  bool found = find_cell(this, false);
+
+  if (settings::run_mode != RUN_MODE_PLOTTING && (!found)) {
+    // If a cell is still not found, there are two possible causes: 1) there is
+    // a void in the model, and 2) the particle hit a surface at a tangent. If
+    // the particle is really traveling tangent to a surface, if we move it
+    // forward a tiny bit it should fix the problem.
+
+    n_coord_ = 1;
+    this->r() += TINY_BIT * this->u();
+
+    // Couldn't find next cell anywhere! This probably means there is an actual
+    // undefined region in the geometry.
+
+    if (!find_cell(this, false)) {
+      this->mark_as_lost("After particle " + std::to_string(id_) +
+        " crossed surface " + std::to_string(surf->id_) +
+        " it could not be located in any cell and it did not leak.");
+      return;
+    }
+  }
+}
+
+void
+Particle::mark_as_lost(const char* message)
+{
+  // Print warning and write lost particle file
+  warning(message);
```

Comment:
I see that using message names for "warning" and "printing" are being used. This is not right. Use a C++ logger or capture all write events in a custom class. This will not scale.